src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -394,4 +355,145 @@ private[concurrent] final object Promise {
+  final class Callback[T](
+    final val executor: ExecutionContext,
+    final val onComplete: Try[T] => Any) extends Callbacks[T] with Runnable with OnCompleteRunnable {
+
+    private[this] final var value: AnyRef = executor // value is initially the EC
+
+    override final def run(): Unit = value match {
+      case null => // already executed once already
+      case v: Try[T] =>
+        value = null
+        try onComplete(v) catch { case NonFatal(e) => executor reportFailure e }
+      case e if e eq executor =>
+        throw new IllegalStateException("Callback value must be set when running")
+    }
+
+    override final def submitWithValue(v: Try[T]): Unit = 
+      if (this ne NoopCallback) {
+        val e = value
+        if (e eq executor) {
+          value = v // Safe publication of `value`, to run(), is achieved via `executor.execute(this)`
+          // Note that we cannot prepare the ExecutionContext at this point, since we might already be running on a different thread!            
+          try executor.execute(this) catch { case NonFatal(t) => executor reportFailure t }
+        } // else … already submitted or already executed
+      }
+
+    override final def prepend[U >: T](c: Callbacks[U]): Callbacks[U] = c match {
+      case a: Callback[U] =>
+        if (a eq NoopCallback) this
+        else if (this eq NoopCallback) a
+        else if (this eq a) this
+        else ManyCallbacks.two(this, a)
+      case m: ManyCallbacks[U] => m append this // m append this == this prepend m
+    }
+
+    override def toString: String = s"Callback($executor, $onComplete)"
+  }
+
+  object ManyCallbacks {
+    def two[U](second: Callbacks[U], first: Callbacks[U]): ManyCallbacks[U] =
+      new ManyCallbacks[U](NoopCallback, NoopCallback, second, first, 2)
+    def three[U](third: Callbacks[U], second: Callbacks[U], first: Callbacks[U]): ManyCallbacks[U] =
+      new ManyCallbacks[U](NoopCallback, third, second, first, 1)
+    def four[U](fourth: Callbacks[U], third: Callbacks[U], second: Callbacks[U], first: Callbacks[U]): ManyCallbacks[U] =
+      new ManyCallbacks[U](fourth, third, second, first, 0)
+  }
+
+  final class ManyCallbacks[+T] private[ManyCallbacks] (
+    final val c1: Callbacks[T],
+    final val c2: Callbacks[T],
+    final val c3: Callbacks[T],
+    final val c4: Callbacks[T],
+    private[ManyCallbacks] final val remainingCapacity: Int) extends Callbacks[T] {
+
+    //Don't want to incur the runtime ovehead of these checks, but this invariant will hold true:
+    //require(c3 ne NoopCallback)
+    //require(c4 ne NoopCallback)
+
+    private[this] final def merge0[U >: T](m: ManyCallbacks[U]): ManyCallbacks[U] =
+      (m.remainingCapacity: @switch) match {
+        case 0 => ManyCallbacks.two(this, m)
+        case 1 => ManyCallbacks.four(c1, c2, c3, ManyCallbacks.four(c4, m.c2, m.c3, m.c4))
+        case 2 => ManyCallbacks.three(c1, c2, ManyCallbacks.four(c3, c4, m.c3, m.c4))
```

Comment:
Is it better to first fill up the higher levels of the trie before the lower levels or the other way around?
Currently it seems to me that turning things around here (i.e. nest a 3-element instance in a 4-element one) would be slightly better in certain cases.

Reply:
Great question. I'm still a bit of the fence as to whether this structure pays for itself or not.